### PR TITLE
integrate Sentry into telemetry stack

### PR DESCRIPTION
**Pull request by pH14**
Bit of a skunkworks project, mostly to start a conversation -- this PR integrates Sentry into envd/storaged/computed. It adds in an optional `SENTRY_DSN` param that, when set, will send `error!` logs and and `panic`s to Sentry. Seems like it could help us keep better track of the crashes we see in staging/prod

It builds off of https://github.com/MaterializeInc/materialize/pull/14969 so until that PR is merged, the diff shows a bit more than it needs to. The last commit has the most relevant changes.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by benesch** [comment]
Oh, hallelujah! Thank you so much, @pH14! I've wanted this forever. It's loosely tracked in https://github.com/MaterializeInc/cloud/issues/3738.

**Comment by benesch** [comment]
I'll take a look when https://github.com/MaterializeInc/materialize/pull/14969 merges.

**Comment by pH14** [comment]
I've tried this locally and it seems to do what it ought to. I'll branch deploy tomorrow/Thursday after the current release goes out to verify it looks right in a real cluster!

File: src/ore/Cargo.toml [hunk 1]
```diff
@@ -31,6 +32,7 @@ pin-project = "1.0.12"
 prometheus = { version = "0.13.2", default-features = false, optional = true }
 smallvec = { version = "1.9.0", optional = true }
 stacker = { version = "0.1.15", optional = true }
+sentry = "0.27.0"
```
Comment (benesch) [inline src/ore/Cargo.toml:35]:
```suggestion
sentry = { version = "0.27.0", optional = true }
```

File: src/ore/Cargo.toml [hunk 2]
```diff
@@ -59,6 +61,7 @@ hyper-tls = { version = "0.5.0", optional = true }
 opentelemetry = { git = "https://github.com/MaterializeInc/opentelemetry-rust.git", features = ["rt-tokio", "trace"], optional = true }
 opentelemetry-otlp = { git = "https://github.com/MaterializeInc/opentelemetry-rust.git", optional = true }
 console-subscriber = { version = "0.1.8", optional = true }
+sentry-tracing = "0.27.0"
```
Comment (benesch) [inline src/ore/Cargo.toml:64]:
```suggestion
{ sentry-tracing = "0.27.0", optional = true }
```

File: src/ore/Cargo.toml [hunk 3]
```diff
@@ -59,6 +61,7 @@ hyper-tls = { version = "0.5.0", optional = true }
...
 
 [dev-dependencies]
 tokio = { version = "1.19.2", features = ["macros"] }
```
Comment (benesch) [inline src/ore/Cargo.toml:67]: Can't leave a line comment on it, but you'll need to add `sentry` and `sentry-tracing` to the `tracing_` feature.

File: src/ore/src/tracing/mod.rs [hunk 4]
```diff
@@ -66,6 +68,17 @@ pub struct TracingConfig {
     #[cfg_attr(nightly_doc_features, doc(cfg(feature = "tokio-console")))]
     #[cfg(feature = "tokio-console")]
     pub tokio_console: Option<TokioConsoleConfig>,
+    /// Optional Sentry configuration
```
Comment (benesch) [inline src/ore/src/tracing/mod.rs:71]:
```suggestion
    /// Optional Sentry configuration.
```

File: src/ore/src/tracing/mod.rs [hunk 5]
```diff
@@ -66,6 +68,17 @@ pub struct TracingConfig {
...
+    pub sentry: Option<SentryConfig>,
+}
+
+/// Configures Sentry reporting
```
Comment (benesch) [inline src/ore/src/tracing/mod.rs:75]:
```suggestion
/// Configures Sentry reporting.
```

File: src/ore/src/tracing/mod.rs [hunk 6]
```diff
@@ -66,6 +68,17 @@ pub struct TracingConfig {
...
+#[derive(Debug, Clone)]
+pub struct SentryConfig {
+    /// Sentry data source name to submit events to
```
Comment (benesch) [inline src/ore/src/tracing/mod.rs:78]:
```suggestion
    /// Sentry data source name to submit events to.
```

File: src/ore/src/tracing/mod.rs [hunk 7]
```diff
@@ -66,6 +68,17 @@ pub struct TracingConfig {
...
+    pub dsn: String,
+    /// Additional tags to include on each Sentry event/exception
```
Comment (benesch) [inline src/ore/src/tracing/mod.rs:80]:
```suggestion
    /// Additional tags to include on each Sentry event/exception.
```

File: src/orchestrator-tracing/src/lib.rs [hunk 8]
```diff
@@ -174,6 +174,9 @@ pub struct TracingCliArgs {
         default_value = "1h",
     )]
     pub tokio_console_retention: Duration,
+    /// Sentry DSN to submit events, exceptions (e.g. panics) to
```
Comment (benesch) [inline src/orchestrator-tracing/src/lib.rs:177]:
```suggestion
    /// Sentry data source to submit events and exceptions (e.g. panics) to.
```

File: src/ore/Cargo.toml [hunk 9]
```diff
@@ -22,6 +22,7 @@ clap = { version = "3.2.20", features = ["env"], optional = true }
 ctor = { version = "0.1.23", optional = true }
 either = "1.8.0"
 futures = { version = "0.3.24", optional = true }
+mz-build-info = { path = "../build-info" }
```
Comment (benesch) [inline src/ore/Cargo.toml:25]: ~Aw, hm, a bit of a bummer to make ore dependent on `mz-build-info`, as I try to keep `mz-ore` relatively free of Materialize-specific stuff, and `mz-build-info` has some Materialize-specific stuff.~ Seems worth doing, though; anything else would be real painful.

**Review by benesch — APPROVED** [review]
Thank you much, @pH14! Can you file an issue against cloud (or a PR!) to plumb the DSN through as an env var or CLI flag before you merge this, so that we don't forget?

File: src/ore/Cargo.toml (hunk 9 above)
Comment (pH14) [reply src/ore/Cargo.toml:25]: oh I think we can work around this without too much pain. we can just pass in a string or two for the build info

**Pull request merged**